Sources/SentryCrash/Recording/Tools/SentryHook.c

```diff
@@ -56,6 +28,80 @@ sentrycrash__async_backtrace_decref(sentrycrash_async_backtrace_t *bt)
     }
 }
 
+/**
+ * This is a poor-mans concurrent hashtable.
+ * We have N slots, using modulo of the thread ID. Using atomic load / compare-exchange to make sure
+ * that the slot indeed belongs to the thread we want to work with.
+ */
+
+#define SENTRY_MAX_ASYNC_THREADS (128 - 1)
+
+typedef struct {
+    SentryCrashThread thread;
+    sentrycrash_async_backtrace_t *backtrace;
+} sentrycrash_async_caller_t;
+
+static sentrycrash_async_caller_t sentry_async_callers[SENTRY_MAX_ASYNC_THREADS] = { 0 };
+
+static size_t
+sentrycrash__thread_idx(SentryCrashThread thread)
+{
+    return (thread * 0x100000001b3 + 0xcbf29ce484222325) % SENTRY_MAX_ASYNC_THREADS;
+}
+
+sentrycrash_async_backtrace_t *
+sentrycrash_get_async_caller_for_thread(SentryCrashThread thread)
+{
+    if (!__atomic_load_n(&hooks_active, __ATOMIC_RELAXED)) {
+        return NULL;
+    }
+
+    size_t idx = sentrycrash__thread_idx(thread);
+    sentrycrash_async_caller_t *caller = &sentry_async_callers[idx];
+    if (__atomic_load_n(&caller->thread, __ATOMIC_ACQUIRE) == thread) {
+        sentrycrash_async_backtrace_t *backtrace = caller->backtrace;
```

Comment:
i don't really see how this read is safe, even in conjunction with the `sentrycrash__set_async_caller`.  AFAIK both of these can be called concurrently from many threads, and I don't think anything protects this read in case of a hash collision.